### PR TITLE
Use user stream by default in copy_to_external/feed_ndarray

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

#### Why we need this PR?
*Pick one, remove the rest*
- It adds new feature needed to have the possibility to use DALI's internal user stream by default (instead of using default CUDA stream) while keeping the possibility to specify the stream manually

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     *Change default CUDA stream to `py::none` which means "use the internal user stream". The user can always specify cuda_stream=0 explicitly to use the CUDA default stream, or specify any other stream*
 - Affected modules and functionalities:
     *copy_to_external*
 - Key points relevant for the review:
     *Everything*
 - Validation and testing:
     *Existing tests*
 - Documentation (including examples):
     *Doc str updated*

**JIRA TASK**: *[Use DALI-XXXX or NA]*

### CODE REVIEW

**Comment by jantonguirao** [comment]
!build

File: dali/python/nvidia/dali/plugin/paddle.py [hunk 1]
```diff
@@ -70,9 +70,15 @@ def feed_ndarray(dali_tensor, ptr):
                     Tensor from which to copy
     `ptr` : LoDTensor data pointer
             Destination of the copy
+    `cuda_stream` : Any value that can be casted to cudaStream_t
+                    CUDA stream to be used for the copy
+                    (if not provided, an internal user stream will be selected)
     """
     c_type_pointer = ctypes.c_void_p(ptr)
-    dali_tensor.copy_to_external(c_type_pointer)
+    if isinstance(dali_tensor, nvidia.dali.backend.TensorGPU):
```
Comment (JanuszL) [inline dali/python/nvidia/dali/plugin/paddle.py:78]: What about TensorListGPU?

File: dali/python/nvidia/dali/plugin/mxnet.py [hunk 2]
```diff
@@ -57,7 +60,10 @@ def feed_ndarray(dali_tensor, arr):
     ptr = ctypes.c_void_p()
     mx.base._LIB.MXNDArrayGetData(arr.handle, ctypes.byref(ptr))
     # Copy data from DALI tensor to ptr
-    dali_tensor.copy_to_external(ptr)
+    if isinstance(dali_tensor, nvidia.dali.backend.TensorGPU):
```
Comment (JanuszL) [inline dali/python/nvidia/dali/plugin/mxnet.py:63]: I wonder how `mx.nd.zeros` works and if we won't have the same problem as in PyTorch.

File: dali/python/nvidia/dali/plugin/paddle.py (hunk 1 above)
Comment (jantonguirao) [reply dali/python/nvidia/dali/plugin/paddle.py:78]: right, missed that

File: dali/python/nvidia/dali/plugin/mxnet.py (hunk 2 above)
Comment (jantonguirao) [reply dali/python/nvidia/dali/plugin/mxnet.py:63]: There is this call to `mx.base._LIB.MXNDArrayWaitToWrite(arr.handle)`, not sure if that'd be enough

Comment (JanuszL) [reply dali/python/nvidia/dali/plugin/mxnet.py:63]: Not to mention PaddlePaddle.

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

File: dali/python/nvidia/dali/plugin/mxnet.py [hunk 3]
```diff
@@ -37,7 +38,7 @@ def _wait_to_write(arr):
         raise RuntimeError("Can only wait for NDArray")
     mx.base._LIB.MXNDArrayWaitToWrite(arr.handle)
 
-def feed_ndarray(dali_tensor, arr):
+def feed_ndarray(dali_tensor, arr, cuda_stream = None):
```
Comment (JanuszL) [inline dali/python/nvidia/dali/plugin/mxnet.py:41]: I would add some comment that in the MXNet in most cases None and 0 are sane. Other values doesn't make much sense unless you create this stream on your own somehow.

**Comment by jantonguirao** [comment]
!build

File: dali/python/nvidia/dali/plugin/mxnet.py (hunk 3 above)
Comment (jantonguirao) [reply dali/python/nvidia/dali/plugin/mxnet.py:41]: done

**Comment by jantonguirao** [comment]
!build

File: dali/test/python/test_fw_iterators.py [hunk 4]
```diff
@@ -428,6 +428,70 @@ def test_pytorch_iterator_not_fill_last_batch_pad_last_batch():
...
+                                            std=[0.229 * 255,0.224 * 255,0.225 * 255])
+
+    def define_graph(self):
+        jpegs, _ = self.input(name="Reader")
+        images = self.decode_gpu(jpegs)
+        images = self.res(images)
+        output = self.cmnp(images)
+        return output
+
+def test_pytorch_iterator_feed_ndarray():
```
Comment (JanuszL) [inline dali/test/python/test_fw_iterators.py:453]: Is it possible to add PaddlePaddle test as well?

Comment (jantonguirao) [reply dali/test/python/test_fw_iterators.py:453]: will do

**Comment by jantonguirao** [comment]
!build

**Pull request merged**